guesung/서버로 사진파일과 객체 전송하기 (CORS policy, proxy, MultiPart, Form Data).md

```diff
@@ -0,0 +1,128 @@
+// package.json
+{
+	...
+    "proxy": "http://localhost:8000"
+}
+```
+
+> ### 2. http-proxy-middleware사용하기
+npm으로 http-proxy-middleware를 설치해줍니다.
+src>setupProxy.js라는 파일을 생성 후 다음과 같은 코드를 작성해줍니다.
+```javascript
+// src/setupProxy.jsconst { createProxyMiddleware } = require('http-proxy-middleware');
+module.exports = function(app) {
+  app.use(
+    '/api',
+    createProxyMiddleware({
+      target: 'http://localhost:5000',
+      changeOrigin: true,
+    })
+  );
+};
+```
+그런데, 백엔드 서버측에서 proxy-middleware까지 요구하지 않는 경우, setupProxy.js 파일로 인해 500오류가 계속 발생할 수 있습니다. 
+![](https://velog.velcdn.com/images/gueit214/post/af36ebf8-c0c5-4066-8ba8-e500b03818f8/image.png)
+그런 경우, setupProxy.js파일을 아애 삭제해주셔야 합니다.
+###### 이 파일 하나 때문에 삽질 엄청 했습니다 ㅜㅜ
+
+-----
+위와 같은 proxy설정을 해주었다면 이제 MultiFile과 Formdata를 전송해보도록 하겠습니다.
+>
+### Multipart 
+웹 클라이언트가 요청을 보낼 때, http 프로토콜의 바디 부분에 데이터를 여러 부분으로 나눠서 보내는 것
+###### https://codingnotes.tistory.com/73
+웹 클라이언트가 서버에게 파일을 업로드할 때, http 프로토콜의 바디 부분에 파일정보를 담아서 전송을 하는데, **파일을 한번에 여러개 전송을 하면 body 부분에 파일이 여러개의 부분으로 연결되어 전송**된다. 이렇게 여러 부분으로 나뉘어서 전송되는 것을 Multipart data라고 합니다.
+
+
+> 
+### FormData
+XMLHttpRequest 전송을 위하여 설계된 특수한 객체 형태
+###### https://inpa.tistory.com/entry/JS-%F0%9F%93%9A-FormData-%EC%A0%95%EB%A6%AC-fetch-api
+FormaData는 key, value형식으로 되어있습니다.
+```javascript
+formData = new FormData();
+formData.append('key',value);
+```
+>>
+#### 용도
+- 회원가입을 할 때
+- 파일, 제목, 내용을 DB에 저장하는 페이지가 있을 때
+-> 파일을 http 통신으로 전송할 때 사용
+>
+**값은 무조건 "문자열"로 자동 변환** 되기에 객체나 배열 같은 복잡한 데이터 넣을 수 없습니다. 객체같은 경우, 아래의 코드처럼 for문을 돌려 key와 value를 분리한 후 넣어주면 됩니다. 
+```javascript
+Object.entries(obj).forEach(item => formData.append(item[0], item[1]));
+```
+FormData에 이미지를 담고싶은 경우 아래와 같이 value값에 e.target.files[0]을 넣어주면 됩니다.
+```javascript
+const onUploadImage = useCallback(async e => {
+	const formData = new FormData();
+	formData.append('profiles', e.target.files[0]);
+}, []);
+```
+그 외의 방법으로는 base64, buffer, 2진 data형식으로 서버로 전송할 수도 있습니다.
+###### [참고](https://inpa.tistory.com/entry/JS-%F0%9F%93%9A-Base64-Blob-ArrayBuffer-File-%EB%8B%A4%EB%A3%A8%EA%B8%B0-%EC%A0%95%EB%A7%90-%EC%9D%B4%ED%95%B4%ED%95%98%EA%B8%B0-%EC%89%BD%EA%B2%8C-%EC%84%A4%EB%AA%85)
+
+>
+### Blob
+Blob(Binay Large Object)은 이미지, 사운드, 비디오와 같은 멀티미디어 데이터를 다룰 때 사용
+###### https://heropy.blog/2019/02/28/blob/
+File 객체도 name과 lastModifiedDate 속성이 존재하는 Blob 객체입니다.
+
+>
+제가 진행한 프로젝트의 코드 일부입니다. 제출버튼을 누르면 실행되는 함수입니다.
+`const formData=new FormData();`로 `FormData`를 생성해줍니다. 그리고 FormData는 key와 value값으로 구성되어있으므로 key에는 'profiles', value에는 imageFile을 넣어줍니다. imageFile은 File형식입니다. 그리고 data도 `key`와 `value`로 이루어진 객체입니다. 그렇기에 FormaData의 value값으로 바로 넣을 수 없습니다.
+```javascript
+const data = {
+    name: 'test'
+}
+const submitBtn = async () => {
+	const formData = new FormData();
+	formData.append('profiles', imageFile);
+	for (let [key, value] of Object.entries(data)) {
+		formData.append(key, value);
+	}
+	await axios({
+		method: 'POST',
+		url: `주소/api/user/signup`,
+		headers: {
+			'Content-Type': 'multipart/form-data',
```

Comment:
안녕하세요 규성님

작업환경에서 axios 버전이 궁금합니다! 제가 알기로는 axios 0.27이후 버전에서는
Content-Type이 multipart/form-data 일때 application/x-www-form-urlencoded 으로 변환되어서 갑니다.

해당 이슈는 axios github에서도 논의된적 있습니다.
[v0.27.1 FormData is sent as 'application/x-www-form-urlencoded' instead of 'multipart/form-data'](https://github.com/axios/axios/issues/4627)

Reply:
버전 0.27.2입니다 !! 
앗 정말요 multipart/form-data로 전송할 경우 문제가 있는걸까요 ???

Reply:
이후 글을 읽어보니 0.27 때 hotfix로 고쳐졌고 

하지만 2주전 릴리즈된 1.1.3 에서 또 발생한다고 하네요